File: mwdumper/mwimport/XmlDumpReader.cs
/*
 * MediaWiki import/export processing tools
 * Copyright 2005 by Brion Vibber
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * $Id$
 */

namespace MediaWiki.Import {
	using System;
	using System.Collections;
	using System.IO;
	using System.Text;
	using System.Xml;

	delegate void ElementHandler();

	public class XmlDumpReader {
		XmlTextReader _reader;
		IDumpWriter _writer;
		
		IList _path;
		IDictionary _handlers;
		
		Siteinfo _siteinfo;
		Page _page;
		bool _pageSent;
		Contributor _contrib;
		Revision _rev;
		
		public XmlDumpReader(TextReader inputStream, IDumpWriter writer) {
			_reader = new XmlTextReader(inputStream);
			_reader.WhitespaceHandling = WhitespaceHandling.Significant;
			_writer = writer;
		}
		
		public void ReadDump() {
			_handlers = new Hashtable();
			_handlers["/mediawiki"] = new ElementHandler(ReadMediaWiki);
			_handlers["/mediawiki/siteinfo"] = new ElementHandler(ReadSiteInfo);
			_handlers["/mediawiki/siteinfo/sitename"] = new ElementHandler(ReadSitename);
			_handlers["/mediawiki/siteinfo/base"] = new ElementHandler(ReadBase);
			_handlers["/mediawiki/siteinfo/generator"] = new ElementHandler(ReadGenerator);
			_handlers["/mediawiki/siteinfo/case"] = new ElementHandler(ReadCase);
			_handlers["/mediawiki/siteinfo/namespaces"] = new ElementHandler(ReadNamespaces);
			_handlers["/mediawiki/siteinfo/namespaces/namespace"] = new ElementHandler(ReadNamespace);
			_handlers["/mediawiki/page"] = new ElementHandler(ReadPage);
			_handlers["/mediawiki/page/title"] = new ElementHandler(ReadTitle);
			_handlers["/mediawiki/page/id"] = new ElementHandler(ReadPageId);
			_handlers["/mediawiki/page/restrictions"] = new ElementHandler(ReadRestrictions);
			_handlers["/mediawiki/page/revision"] = new ElementHandler(ReadRevision);
			_handlers["/mediawiki/page/revision/id"] = new ElementHandler(ReadRevId);
			_handlers["/mediawiki/page/revision/timestamp"] = new ElementHandler(ReadTimestamp);
			_handlers["/mediawiki/page/revision/contributor"] = new ElementHandler(ReadContributor);
			_handlers["/mediawiki/page/revision/contributor/username"] = new ElementHandler(ReadUsername);
			_handlers["/mediawiki/page/revision/contributor/id"] = new ElementHandler(ReadContribId);
			_handlers["/mediawiki/page/revision/contributor/ip"] = new ElementHandler(ReadIp);
			_handlers["/mediawiki/page/revision/comment"] = new ElementHandler(ReadComment);
			_handlers["/mediawiki/page/revision/minor"] = new ElementHandler(ReadMinor);
			_handlers["/mediawiki/page/revision/text"] = new ElementHandler(ReadText);
			
			_path = new ArrayList();
			ReadThrough();
			
			_writer.Close();
			_reader.Close();
		}
		
		private void ReadThrough() {
			string current = _reader.LocalName;
			
			while (_reader.Read()) {
				string name = _reader.LocalName;
				if (_reader.NodeType == XmlNodeType.Element) {
					//Console.WriteLine("<!-- open: " + name + "-->");
					_path.Add(name);
					string pathKey = PathKey();
					//Console.WriteLine("<!-- path: " + pathKey + "-->");
					if (_handlers.Contains(pathKey)) {
						ElementHandler handler = (ElementHandler)_handlers[pathKey];
						handler();
						_path.RemoveAt(_path.Count - 1);
						continue;
					} else {
						throw new ArgumentException("Unexpected element, path " + pathKey);
					}
				} else if (_reader.NodeType == XmlNodeType.EndElement && current == name) {
					return;
				}
			}
		}
		
		private string PathKey() {
			StringBuilder str = new StringBuilder();
			foreach (string node in _path) {
				str.Append('/');
				str.Append(node);
			}
			return str.ToString();
		}
		
		private string ReadElementContent() {
			StringBuilder val = new StringBuilder();
			if (_reader.IsEmptyElement) {
				return "";
			}
			while (_reader.Read()) {
				switch (_reader.NodeType) {
				case XmlNodeType.SignificantWhitespace:
				case XmlNodeType.Text:
					val.Append(_reader.Value);
					break;
				case XmlNodeType.EndElement:
					return val.ToString();
				default:
					// ignore
					break;
				}
			}
			return val.ToString();
		}
		
		// ----------
		
		private void ReadMediaWiki() {
			_siteinfo = null;
			_writer.WriteStartWiki();
			ReadThrough();
			_writer.WriteEndWiki();
			_siteinfo = null;
		}
		
		// ------------------
		
		private void ReadSiteInfo() {
			_siteinfo = new Siteinfo();
			ReadThrough();
			_writer.WriteSiteinfo(_siteinfo);
		}
		
		private void ReadSitename() {
			_siteinfo.Sitename = ReadElementContent();
		}
		
		private void ReadBase() {
			_siteinfo.Base = ReadElementContent();
		}
		
		private void ReadGenerator() {
			_siteinfo.Generator = ReadElementContent();
		}
		
		private void ReadCase() {
			_siteinfo.Case = ReadElementContent();
		}
		
		private void ReadNamespaces() {
			_siteinfo.Namespaces = new Hashtable();
			ReadThrough();
		}
		
		private void ReadNamespace() {
			int key = XmlConvert.ToInt32(_reader.GetAttribute("key"));
			_siteinfo.Namespaces[key] = ReadElementContent();
		}
		
		// -----------
		
		private void ReadPage() {
			_page = new Page();
			_pageSent = false;
			ReadThrough();
			
			if (_pageSent)
				_writer.WriteEndPage();
			_page = null;
		}
		
		private void ReadTitle() {
			_page.Title = new Title(ReadElementContent(), _siteinfo.Namespaces);
		}
		
		private void ReadPageId() {
			_page.Id = XmlConvert.ToInt32(ReadElementContent());
		}
		
		private void ReadRestrictions() {
			_page.Restrictions = ReadElementContent();
		}
		
		// ------
		
		private void ReadRevision() {
			if (!_pageSent) {
				_writer.WriteStartPage(_page);
				_pageSent = true;
			}
			
			_rev = new Revision();
			ReadThrough();
			_writer.WriteRevision(_rev);
			_rev = null;
		}

		private void ReadRevId() {
			_rev.Id = XmlConvert.ToInt32(ReadElementContent());
		}

		private void ReadTimestamp() {
			_rev.Timestamp = ParseUTCTimestamp(ReadElementContent());
		}

		private void ReadComment() {
			_rev.Comment = ReadElementContent();
		}

		private void ReadMinor() {
			_rev.Minor = true;
			_reader.Skip();
		}

		private void ReadText() {
			_rev.Text = ReadElementContent();
		}
		
		// -----------
		private void ReadContributor() {
			_contrib = null;
			
			ReadThrough();
			
			if (_contrib == null)
				throw new ArgumentException("Invalid contributor");
			
			_rev.Contributor = _contrib;
			_contrib = null;
		}


		private void ReadUsername() {
			_contrib = new Contributor(ReadElementContent());
		}

		private void ReadContribId() {
			_contrib.Id = XmlConvert.ToInt32(ReadElementContent());
		}
		
		private void ReadIp() {
			_contrib = new Contributor(ReadElementContent());
		}
		
		private DateTime ParseUTCTimestamp(string text) {
			// 2003-10-26T04:50:47Z
			//
			// We're doing this manually because:
			// * XmlConvert.ToDateTime() is slow
			// * XmlConvert.ToDateTime() is lossy at local DST boundary
			// * DateTime.ParseExact seems to be similarly lossy, unless
			//   there's a magic formula I haven't found yet.
			string trimmed = text.Trim();
			return new DateTime(
				int.Parse(trimmed.Substring(0,4)),   // year
				int.Parse(trimmed.Substring(5,2)),   // month
				int.Parse(trimmed.Substring(8,2)),   // day
				int.Parse(trimmed.Substring(11,2)),  // hour
				int.Parse(trimmed.Substring(14,2)),  // minute
				int.Parse(trimmed.Substring(17,2))); // second
		}
	}
}

File: mwdumper/src/org/mediawiki/importer/XmlDumpReader.java
/*
 * MediaWiki import/export processing tools
 * Copyright 2005 by Brion Vibber
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * $Id$
 */

package org.mediawiki.importer;

import java.io.IOException;
import java.io.InputStream;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.TimeZone;

import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

public class XmlDumpReader  extends DefaultHandler {
	InputStream input;
	DumpWriter writer;
	
	private StringBuffer buffer;
	
	Siteinfo siteinfo;
	Page page;
	boolean pageSent;
	Contributor contrib;
	Revision rev;
	int nskey;
	
	/**
	 * Initialize a processor for a MediaWiki XML dump stream.
	 * Events are sent to a single DumpWriter output sink, but you
	 * can chain multiple output processors with a MultiWriter.
	 * @param inputStream Stream to read XML from.
	 * @param writer Output sink to send processed events to.
	 */
	public XmlDumpReader(InputStream inputStream, DumpWriter writer) {
		input = inputStream;
		this.writer = writer;
	}
	
	/**
	 * Reads through the entire XML dump on the input stream, sending
	 * events to the DumpWriter as it goes. May throw exceptions on
	 * invalid input or due to problems with the output.
	 * @throws IOException
	 */
	public void readDump() throws IOException {
		try {
			SAXParserFactory factory = SAXParserFactory.newInstance();
			SAXParser parser = factory.newSAXParser();
	
			parser.parse(input, this);
		} catch (ParserConfigurationException e) {
			throw new IOException(e.getMessage());
		} catch (SAXException e) {
			throw new IOException(e.getMessage());
		}
		writer.close();
	}
	
	// --------------------------
	// SAX handler interface methods:
	
	public void startElement(String uri, String localname, String qName, Attributes attributes) throws SAXException {
		// Clear the buffer for character data; we'll initialize it
		// if and when character data arrives -- at that point we
		// have a length.
		buffer = null;
		try {
			// frequent tags:
			if (qName.equals("revision")) openRevision();
			else if (qName.equals("contributor")) openContributor();
			else if (qName.equals("page")) openPage();
			// rare tags:
			else if (qName.equals("mediawiki")) openMediaWiki();
			else if (qName.equals("siteinfo")) openSiteinfo();
			else if (qName.equals("namespaces")) openNamespaces();
			else if (qName.equals("namespace")) openNamespace(attributes);
		} catch (IOException e) {
			throw new SAXException(e);
		}
	}
	
	public void characters(char[] ch, int start, int length) {
		if (buffer == null)
			buffer = new StringBuffer(length);
		buffer.append(ch, start, length);
	}
	
	public void endElement(String uri, String localname, String qName) throws SAXException {
		try {
			// frequent tags:
			if (qName.equals("id")) readId();
			else if (qName.equals("revision")) closeRevision();
			else if (qName.equals("timestamp")) readTimestamp();
			else if (qName.equals("text")) readText();
			else if (qName.equals("contributor")) closeContributor();
			else if (qName.equals("username")) readUsername();
			else if (qName.equals("ip")) readIp();
			else if (qName.equals("comment")) readComment();
			else if (qName.equals("minor")) readMinor();
			else if (qName.equals("page")) closePage();
			else if (qName.equals("title")) readTitle();
			else if (qName.equals("restrictions")) readRestrictions();
			// rare tags:
			else if (qName.equals("mediawiki")) closeMediaWiki();
			else if (qName.equals("siteinfo")) closeSiteinfo();
			else if (qName.equals("sitename")) readSitename();
			else if (qName.equals("base")) readBase();
			else if (qName.equals("generator")) readGenerator();
			else if (qName.equals("case")) readCase();
			else if (qName.equals("namespaces")) closeNamespaces();
			else if (qName.equals("namespace")) closeNamespace();
		} catch (IOException e) {
			throw new SAXException(e);
		}
	}

	// ----------
	
	void openMediaWiki() throws IOException {
		siteinfo = null;
		writer.writeStartWiki();
	}
	
	void closeMediaWiki() throws IOException {
		writer.writeEndWiki();
		siteinfo = null;
	}
	
	// ------------------
		
	void openSiteinfo() {
		siteinfo = new Siteinfo();
	}
	
	void closeSiteinfo() throws IOException {
		writer.writeSiteinfo(siteinfo);
	}

	private String bufferContents() {
		return buffer == null ? "" : buffer.toString();
	}
	
	void readSitename() {
		siteinfo.Sitename = bufferContents();
	}
	
	void readBase() {
		siteinfo.Base = bufferContents();
	}
	
	void readGenerator() {
		siteinfo.Generator = bufferContents();
	}
	
	void readCase() {
		siteinfo.Case = bufferContents();
	}
	
	void openNamespaces() {
		siteinfo.Namespaces = new NamespaceSet();
	}
	
	void openNamespace(Attributes attribs) {
		nskey = Integer.parseInt(attribs.getValue("key"));
	}
	
	void closeNamespace() {
		siteinfo.Namespaces.add(nskey, bufferContents());
	}

	void closeNamespaces() {
		// NOP
	}
	
	// -----------
	
	void openPage() {
		page = new Page();
		pageSent = false;
	}
	
	void closePage() throws IOException {
		if (pageSent)
			writer.writeEndPage();
		page = null;
	}
	
	void readTitle() {
		page.Title = new Title(bufferContents(), siteinfo.Namespaces);
	}
	
	void readId() {
		int id = Integer.parseInt(bufferContents());
		if (contrib != null)
			contrib.Id = id;
		else if (rev != null)
			rev.Id = id;
		else if (page != null)
			page.Id = id;
		else
			throw new IllegalArgumentException("Unexpected <id> outside a <page>, <revision>, or <contributor>");
	}
	
	void readRestrictions() {
		page.Restrictions = bufferContents();
	}
	
	// ------
	
	void openRevision() throws IOException {
		if (!pageSent) {
			writer.writeStartPage(page);
			pageSent = true;
		}
		
		rev = new Revision();
	}
	
	void closeRevision() throws IOException {
		writer.writeRevision(rev);
		rev = null;
	}

	void readTimestamp() {
		rev.Timestamp = parseUTCTimestamp(bufferContents());
	}

	void readComment() {
		rev.Comment = bufferContents();
	}

	void readMinor() {
		rev.Minor = true;
	}

	void readText() {
		rev.Text = bufferContents();
	}
	
	// -----------
	void openContributor() {
		contrib = null;
	}
	
	void closeContributor() {
		if (contrib == null)
			throw new IllegalArgumentException("Invalid contributor");
		
		rev.Contributor = contrib;
		contrib = null;
	}


	void readUsername() {
		contrib = new Contributor(bufferContents());
	}
	
	void readIp() {
		contrib = new Contributor(bufferContents());
	}
	
	private static final TimeZone utc = TimeZone.getTimeZone("UTC");
	private static Calendar parseUTCTimestamp(String text) {
		// 2003-10-26T04:50:47Z
		// We're doing this manually for now, though DateFormatter might work...
		String trimmed = text.trim();
		GregorianCalendar ts = new GregorianCalendar(utc);
		ts.set(
			Integer.parseInt(trimmed.substring(0,0+4)),     // year
			Integer.parseInt(trimmed.substring(5,5+2)) - 1, // month is 0-based!
			Integer.parseInt(trimmed.substring(8,8+2)),     // day
			Integer.parseInt(trimmed.substring(11,11+2)),   // hour
			Integer.parseInt(trimmed.substring(14,14+2)),   // minute
			Integer.parseInt(trimmed.substring(17,17+2)));  // second
		return ts;
	}
}
